### PR TITLE
add motion click application

### PR DESCRIPTION
**Pull request by pratik-prajapati-imgtec**
Signed-off-by: Pratik Prajapati Pratik.Prajapati@imgtec.com

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,23 @@
 # click-apps-for-openwrt
-OpenWrt based click applications
+
+###1) Application to detect motion using motion click
```
Comment (abhijit-mahajani-imgtec) [inline README.md]: Mention the click details in the readme. Add link to mikroe website for the motion click. Also add few lines about mikroBus interface or link to that.
Follow the same format for Readme. Overview, Build instructions, Usage etc.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Project Properties
+####################
+CMAKE_MINIMUM_REQUIRED (VERSION 2.6.2)
+
+SET(PROJECT_NAME click_apps_for_openwrt)
```
Comment (MJNowakowski) [inline CMakeLists.txt]: why is openwrt in the project name? There's nothing openwrt specific about this

File: src/gpio.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    sprintf(path, "/sys/class/gpio/gpio%" PRIu8 "/value", gpio);
+    if ((fd = open(path, O_RDONLY)) < 0)
+    {
+        perror("error: failed to open gpio\n");
+        return -1;
+    }
+
+    memset(&pfd, 0, sizeof(pfd));
+    pfd.fd = fd;
+    pfd.events = POLLPRI;
```
Comment (MJNowakowski) [inline src/gpio.c:162]: why is it done here? nothing uses this

File: src/gpio.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+    lseek(pfd.fd, 0 , SEEK_SET);
+    status = read(fd, buffer, sizeof(buffer));
```
Comment (MJNowakowski) [inline src/gpio.c:165]: what's the purpose of this?

File: src/gpio.c [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        status = poll(&pfd, 1, timeout * 1000);
+        if (status == -1)
+        {
+            perror("poll()");
+            close(fd);
+            return -1;
+        }
+        else if (status == 0)
+        {
+            printf("timeout\n");
```
Comment (MJNowakowski) [inline src/gpio.c]: what's the point of timeout if it does nothing more than just print a timeout string?

File: src/motion_click_app.c [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+#include <unistd.h>
+#include "gpio.h"
+
+#define TIMEOUT  5 /* seconds */
+
+typedef struct
+{
+    uint8_t timeout; /* timeout in seconds */
+    uint8_t gpio;
+} CmdOpts;
```
Comment (MJNowakowski) [inline src/motion_click_app.c]: why CamelCase?

File: src/gpio.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+#include <stdbool.h>
+
+#define IN       0
+#define OUT      1
+#define RISING   0
+#define FALLING  1
+
+#define GPIO_FOR_MIKROBUS(no)  ((no) == 1 ? 21 : ((no) == 2 ? 24 : 0))
+
+extern bool running;
```
Comment (MJNowakowski) [inline src/gpio.h]: don't use globals

File: src/motion_click_app.c [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            case 't':
+                cmd_opts->timeout = strtoul(optarg, NULL, 0);
+                if (cmd_opts->timeout == 0)
+                {
+                    fprintf(stderr, "error: bad timeout\n");
+                    return -1;
+                }
+                break;
+
+            case 'h':
```
Comment (MJNowakowski) [inline src/motion_click_app.c:80]: h is not a getopt arg?

File: src/motion_click_app.c [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        }
+    }
+    return 0;
+}
+
+int main(int argc, char *argv[])
+{
+    int8_t status = 0;
+    CmdOpts cmd_opts;
+    struct sigaction sa;
```
Comment (MJNowakowski) [inline src/motion_click_app.c]: don't leave uninitialised structs

File: src/gpio.c [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    if (status == -1)
+    {
+        fprintf(stderr, "Failed to set gpio edge type\n");
+    }
+
+    close(fd);
+    return status;
+}
+
+int8_t wait_for_interrupt(uint8_t gpio, uint8_t timeout)
```
Comment (MJNowakowski) [inline src/gpio.c]: the whole gpio.c file looks like a set of lib methods that you'd use in other apps, so why do you have motion detector logic embedded in here instead of having a registered client callback?

File: src/gpio.h [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+#define RISING   0
+#define FALLING  1
+
+#define GPIO_FOR_MIKROBUS(no)  ((no) == 1 ? 21 : ((no) == 2 ? 24 : 0))
+
+int8_t gpio_export(uint8_t pin);
+int8_t gpio_unexport(uint8_t pin);
+int8_t gpio_direction(uint8_t pin, uint8_t direction);
+int8_t gpio_edge(uint8_t pin, uint8_t type);
+int8_t wait_for_interrupt(uint8_t gpio, uint8_t timeout);
```
Comment (MJNowakowski) [inline src/gpio.h]: why doesn't this belong to gpio namespace?

File: src/gpio.h [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+ * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#ifndef GPIO_H
+#define GPIO_H
+
+#include <stdbool.h>
+
+#define IN       0
```
Comment (MJNowakowski) [inline src/gpio.h]: these names are too generic for global namespace

File: src/gpio.h [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#ifndef GPIO_H
+#define GPIO_H
+
+#include <stdbool.h>
```
Comment (MJNowakowski) [inline src/gpio.h]: not used anywhere

File: src/motion_click_app.c [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+{
+    running = false;
+    signal(sig, SIG_IGN);
+}
+
+static void print_usage(const char *program)
+{
+    printf("Usage: %s [options]\n\n"
+        " -m  <arg>  mikroBUS number\n"
+        " -t  <arg>  timeout (in seconds) between successive attempts\n"
```
Comment (MJNowakowski) [inline src/motion_click_app.c]: what is the point of this timeout? it doesn't seem to do anything useful

File: src/motion_click_app.c [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+{
+    int opt;
+    uint8_t num;
+
+    opterr = 0;
+
+    while (1)
+    {
+        opt = getopt(argc, argv, "m:t:h");
+        if (opt == -1)
```
Comment (MJNowakowski) [inline src/motion_click_app.c]: change the code to get rid of this extra if - it will make the code far more readable

File: src/motion_click_app.c [hunk 16]
```diff
@@ -56,21 +56,15 @@ static void print_usage(const char *program)
         program);
 }
 
-static int parse_cmd_opts(int argc, char *argv[], uint8_t *gpio)
+static int8_t parse_cmd_opts(int argc, char *argv[], uint8_t *gpio)
```
Comment (MJNowakowski) [inline src/motion_click_app.c]: why did you change this? Sized types are best used only where there is a good reason to do so (interface requirements, hardware width matching, alignment, optimisation). Otherwise it's best to use just standard types as the size is completely irrelevant here.

File: src/motion_click_app.c [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+static void print_usage(const char *program)
+{
+    printf("Usage: %s [options]\n\n"
+        " -m  <arg>  mikroBUS number\n"
+        " -h         display this message\n\n",
+        program);
+}
+
+static int8_t parse_cmd_opts(int argc, char *argv[], uint8_t *gpio)
```
Comment (MJNowakowski) [inline src/motion_click_app.c]: what is the meaning of magic '-1' and '1' return values?

File: src/motion_click_app.c (hunk 16 above)
Comment (pratik-prajapati-imgtec) [inline src/motion_click_app.c]: As I know the return value range here, that's why I took int8_t. But your point makes much sense so I will reverse it.

**Comment by MJNowakowski** [comment]
apart from my last comments, looks OK

File: src/motion_click_app.c (hunk 17 above)
Comment (pratik-prajapati-imgtec) [inline src/motion_click_app.c]: I have used '1' to differentiate between 'h' and other valid options.
well.. using exit(0) instead of return '1' for 'h' option can solve this

Comment (MJNowakowski) [inline src/motion_click_app.c]: But you're not checking the return value for '1' anyway?

File: src/motion_click_app.c (hunk 6 above)
Comment (abhijit-mahajani-imgtec) [inline src/motion_click_app.c]: We have been following CamelCase for structures and its documented in 
https://github.com/CreatorKit/creator-docs/pull/25/files#diff-f09ffbd07117a45e24cdd492cf7ab284 
so we need to decide whether to keep consistency between click apps and CreatorKit apps and use these guidelines or we can define altogether new guidelines or point to linux kernel guidelines.
@Duncan-Frazer-imgtec , @MarcinNowakowski-IMG, @seank-img should we use CreatorKit coding guidelines or linux kernel guidelines for such click applications?

File: README.md [hunk 18]
```diff
@@ -1,2 +1,27 @@
...
+
+###Application to detect motion using [motion click](http://www.mikroe.com/click/motion/)
+
+After starting the app move your hand above the motion click.
+
+Press ctrl+c to exit the application.
+
+Two options can be given:
+
+1) mikroBUS number to specify where motion click is sitting.
```
Comment (abhijit-mahajani-imgtec) [inline README.md]: sitting -> connected?

File: src/gpio.c [hunk 19]
```diff
@@ -142,11 +142,11 @@ int8_t gpio_edge(uint8_t pin, uint8_t type)
     return status;
 }
 
-int8_t gpio_wait_for_interrupt(uint8_t gpio, uint8_t timeout)
+int8_t GpioWaitForInterrupt(uint8_t gpio, uint8_t timeout)
 {
     int fd;
     int8_t status;
-    struct pollfd pfd;
+    struct pollfd Pfd;
```
Comment (abhijit-mahajani-imgtec) [inline src/gpio.c]: first letter small and next capital or such small variable name everything can be small

**Comment by abhijit-mahajani-imgtec** [comment]
Approved.

File: README.md [hunk 20]
```diff
@@ -1,2 +1,27 @@
...
+
+Press ctrl+c to exit the application.
+
+Two options can be given:
+
+1) mikroBUS number to specify where motion click is connected.
+
+2) timeout to specify time between successive attempts.
+
+Default values for [mikroBUS](http://www.mikroe.com/mikrobus/) is 1 and for timeout is 5 seconds
```
Comment (abhijit-mahajani-imgtec) [inline README.md]: You are not accepting the timeout argument, then change the readme accordingly..

File: README.md [hunk 21]
```diff
@@ -1,2 +1,27 @@
 # click-apps-for-openwrt
-OpenWrt based click applications
+
+###Overview
+
```
Comment (abhijit-mahajani-imgtec) [inline README.md:4]: There has to be some build instructions as well. such as to add the feed from openwrt-feeds and then select in the menuconfig etc and then build openwrt.

File: README.md [hunk 22]
```diff
@@ -4,6 +4,12 @@
 
 Applications to show how mikro-e [sensor clicks](http://www.mikroe.com/add-on-boards/sensors/) can be used with Ci40.
 
+###How to install applications of this package in OpenWrt's filesystem
```
Comment (abhijit-mahajani-imgtec) [inline README.md:7]: Just selecting in menuconfig it won't install. Mention that need to build openwrt with usual way, and load the image. Just point to IMGCreator/openwrt readme appropriate sections.

**Comment by abhijit-mahajani-imgtec** [comment]
Approved

**Comment by abhijit-mahajani-imgtec** [comment]
please shift the PR to dev
